rp2040-hal/src/multicore.rs

```diff
@@ -0,0 +1,211 @@
+//! Multicore support
+// See [Chapter ?? Section ??](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
+
+use crate::pac;
+
+#[cfg(feature = "alloc")]
+extern crate alloc;
+
+/// Errors for multicore operations.
+#[derive(Debug)]
+pub enum Error {
+    /// Operation is invalid on this core.
+    InvalidCore,
+    /// Core was unresposive to commands.
+    Unresponsive,
+}
+
+// We pass data to cores via the stack, so we read
+// the data off the stack and into parameters that
+// rust can read here. Ideally this would be a
+// #[naked] function but that is not stable yet.
+static MULTICORE_TRAMPOLINE: [u16; 2] = [
+    0xbd03, // pop {r0, r1, pc} - call wrapper (pc) with r0 and r1
+    0x46c0, // nop - pad this out to 32 bits long
+];
+static mut CORE1_STACK: [usize; 4096] = [0; 4096];
+
+#[inline(always)]
+fn install_stack_guard(stack_bottom: *mut usize) {
+    let core = unsafe { pac::CorePeripherals::steal() };
+
+    // Trap if MPU is already configured
+    if core.MPU.ctrl.read() != 0 {
+        cortex_m::asm::udf();
+    }
+
+    // The minimum we can protect is 32 bytes on a 32 byte boundary, so round up which will
+    // just shorten the valid stack range a tad.
+    let addr = (stack_bottom as u32 + 31) & !31;
+    // Mask is 1 bit per 32 bytes of the 256 byte range... clear the bit for the segment we want
+    let subregion_select = 0xff ^ (1 << ((addr >> 5) & 7));
+    unsafe {
+        core.MPU.ctrl.write(5); // enable mpu with background default map
+        core.MPU.rbar.write((addr & !0xff) | 0x8);
+        core.MPU.rasr.write(
+            1 // enable region
+               | (0x7 << 1) // size 2^(7 + 1) = 256
+               | (subregion_select << 8)
+               | 0x10000000, // XN = disable instruction fetch; no other bits means no permissions
+        );
+    }
+}
+
+#[inline(always)]
+fn core1_setup(stack_bottom: *mut usize) {
+    install_stack_guard(stack_bottom);
+    // TODO: irq priorities
+}
+
+/// Multicore execution management.
+pub struct Multicore<'p> {
+    cores: [Core<'p>; 2],
+}
+
+impl<'p> Multicore<'p> {
+    /// Create a new |Multicore| instance.
+    pub fn new(psm: &'p mut pac::PSM, ppb: &'p mut pac::PPB, sio: &'p mut crate::sio::Sio) -> Self {
+        Self {
+            cores: [
+                Core { inner: None },
+                Core {
+                    inner: Some((psm, ppb, sio)),
+                },
+            ],
+        }
+    }
+
+    /// Get the available |Core| instances.
+    pub fn cores(&mut self) -> &'p mut [Core] {
```

Comment:
Since on RP2040 we always start execution on core 0 and only ever spawn new tasks on core 1 with the current HAL, we should make that path more ergonomic.
If we can use the existing code to get there (so that it stays tested) that would be best, so we get easy to use and can deal with more cores in the future.